[java] Fix AvoidUsingOctalValues false-positive

Gold856 opened the pull request:
## Describe the PR

ASTNumericLiteral did not correctly determine the base when a number ended with a non-number (L, F, D, .) This PR checks if the number ends with one of those characters, and correctly returns a base of 10.

## Related issues

- Fixes #5007

## Ready?

<!-- If you feel like you can help to check off the following tasks, that'd be great. If not, don't worry - we will take care of it. -->

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

jsotuyod commented on pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTNumericLiteral.java:
```diff
@@ -124,18 +124,19 @@ public int getBase() {
-                return 10;
-            default:
-                return 8;
+        switch (image.charAt(image.length() - 1)) {
```
this should have the same length check we have in line 131 for defensiveness

oowekyala commented on pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTNumericLiteral.java:
```diff
@@ -124,18 +124,19 @@ public int getBase() {
-            default:
-                return 8;
+        switch (image.charAt(image.length() - 1)) {
+        case 'l', 'L', 'f', 'F', 'd', 'D', '.':
```
Numeral type suffixes only influence the type and not the base of the number. You can write `0x0123L` and that is a long but still base 16

Gold856 commented:
I added some test cases and fixed the logic; the code now checks for a period in the literal along with a float or double suffix at the end, since those do cause the base to change to base 10.

pmd-test commented:
<!--
  0 Errors
  0 Warnings
  1 Message: Compared to master:<br />
This...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 63 new violations, 0 new errors and 0 new configuration errors,<br />
removes 2 violations, 0 errors and 0 configuration errors.<br />
<a href="https://github.com/pmd/pmd/actions/runs/9102279270?pr=5020">Download full report as build artifact</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

jsotuyod commented on pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTNumericLiteral.java:
```diff
@@ -125,17 +125,26 @@ public int getBase() {
 
     static int getBase(Chars image) {
         if (image.length() > 1 && image.charAt(0) == '0') {
+            if (image.indexOf('.') != -1) {
+                return 10;
+            }
```
this is actually not true…

The value `0x1.0p+2f` is an hexadecimal float literal in Java. It's still base 16.

I'm seeing a lot of false positives in the `UseUnderscoresInNumericLiterals` rule because of this changeset.

Gold856 replied:
Moved this check to after the binary and hex checks.

jsotuyod commented:
Let's see what the regression tester shows, but it seems better.

I'd probably just add a test case before merging on the case `0x1.0000_ffff_eeeep+2f` for the rule `UseUnderscoresInNumericLiterals`, as this was detected by the regression tester but not our own rules, so we are obviously missing this scenario there.

But other than that, if @pmd-test doesn't show anything weird, this is good to go.

Gold856 commented:
Added the float literal test case. Also added code to check for exponents (those can also change an octal to decimal) with a test case.

oowekyala commented on pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTNumericLiteral.java:
```diff
@@ -132,10 +132,19 @@ static int getBase(Chars image) {
-                return 10;
             default:
-                return 8;
+                if (image.indexOf('.') != -1 || image.indexOf('e') != -1 || image.indexOf('E') != -1) {
```
I think instead of all this logic we could add a parameter `boolean isIntegral` to `getBase(Chars)`. Then in the default case we just have `return isIntegral? 8 : 10;`. In `parseIntegralValue` this value is true, we know it's integral. In `getBase()` we can just use the field `isIntegral`. This would simplify this function a lot.

jsotuyod reviewed (approved):
This looks great! Thanks!
